### DISCUSSION TITLE
How to register dynamic entities at runtime?

### DISCUSSION THREAD

**Discussion by ZakRabe** [mikro-orm/mikro-orm#6026] [Q&A] [answered]
I am using postgres. and NestJS.
I am trying to use multiple tables with the same data shape, but different table names. The table names are created dynamically by a user.

I've been unable to figure out how to dynamically register entities at runtime.

My code needs to be able to create new copies of an existing table, and is currently creating them using `raw`. Maybe there is a better way to do this? I assumed not, since the standard way to manage tables is with migrations.

```this.em.raw( `CREATE TABLE ${cloneName} AS SELECT * FROM source_table where 1=0` );```

To create my dynamic entity class, im just using a function that returns a standard decorated class. But maybe these decorators don't work the same way when executed this way inside a function?
```
export const createVersionEntity= (title: string) => {
  @Entity({ tableName: title })
  class VersionEntity {
    constructor(id: string, price: number) {
      this.id = id;
      this.price = price;
    }
    @PrimaryKey()
    id: string;
    @Property({ nullable: true })
    price: number;
  }
  return VersionEntity;
};
```

But when I try to save rows into this cloned table, mikro-orm throws a `ValidationError`:
```ValidationError: Trying to persist not discovered entity of type object.```

This leads me to believe the ORM isn't aware of the entity, which makes sense. I just don't know how to properly register the entity in the ORM

I've tried using the `MetadataStorage`, by cloning the source table metadata, modifying it, and setting it back into the `MetadataStorage`
```this.em.getMetadata().set(cloneName, clonedMetadata)```

I've tried `MikroORM.discoverEntity`, but this doesn't seem to work for me either, I get an error in that method as well. It seems like MikroORM cant get the primary key for the entity?
```
if (meta.properties[pk].reference !== enums_1.ReferenceType.SCALAR) 
....
TypeError: Cannot read properties of undefined (reading 'reference')
```

I've also looked into EntityGenerator, and a dynamic EntityRepository class neither has worked for me. I cant seem to find any documentation on registering entitles at runtime, instead of using startup entity discovery. 

The only detail that is different between these Entities is the tableName, Does anyone have any ideas on how I could solve this problem in a way that mikro-orm supports? It feels like im fighting the library on this, all other features i've used so far are really fantastic and user friendly.

**Comment by B4nan** [comment]
I would use `EntitySchema` instead of decorators, those have global metadata storage and you can't use a class with the same name multiple times for that.

**Reply by ZakRabe** [reply]
> I would use `EntitySchema` instead of decorators, those have global metadata storage and you can't use a class with the same name multiple times for that.

Thanks for the response! Could you please give me an example of how to register the entity I create with `EntitySchema`?

I wrote  a helper to create the entity schema at runtime: 

```
export const generateCardPriceEntity = (versionTitle: string) => {
  const schema = new EntitySchema({
    class: CardPrice,
    tableName: versionTitle,
    name: versionTitle,
    properties: {
      id: { primary: true, type: 'string' },
      price: { type: 'number', nullable: true },
    },
  });
  return schema;
};
```
But I get an error that the ORM cannot find the metadata for the entity. It is unclear what I am expected to do with this EntitySchema object after creating it at runtime


Passing a dynamic class to discoverEntity seems to avoid this, However my code that is trying to create the entities seems to error in the postgres layer when persisting the entities.

```
    const { schema, EntityClass } = generateCardPriceEntity(versionTitle);
    await this.orm.discoverEntity(EntityClass);

    for (const entity of priceData) {
      await this.em.create(
        versionTitle,
        new EntityClass(entity.id, entity.price),
      );
    }
```

```
@mikro-orm\postgresql\PostgreSqlPlatform.js:152
        return `${quote}${id.replace('.', `${quote}.${quote}`)}${quote}`;
                             ^

TypeError: Cannot read properties of undefined (reading 'replace')
    at PostgreSqlPlatform.quoteIdentifier (node_modules\@mikro-orm\postgresql\PostgreSqlPlatform.js:152:30)
```


Should I be passing the base class whenever I create a schema? or should I still be using my function to generate a new class for each uniquely named copy of the table?

After I get this working, I'll submit a PR to update the [docs for EntitySchema](https://mikro-orm.io/docs/entity-schema), and add some examples around this kind of usage

**Reply by B4nan** [reply]
>     class: CardPrice,

You can't use the same class for multiple entities.

>     name: versionTitle,

This will be ignored because you provided `class` option and its name will be used.

If you drop the `class` option, a dynamic class implementation will be created with the provided `name`. You can then use this name in `em.create/find/...` and other methods.

And yes, you should definitely go through `discoverEntity` method.

**Reply by ZakRabe** [reply]
Ok I have made those changes, but I am still not doing anything with the created EntitySchema. 

Here are the pieces of my code im trying to use for this feature

Here is the "base" entity, the shared model I need all these tables to share
```
@Entity()
export class CardPrice {
  @PrimaryKey()
  id: string;
  @Property({ nullable: true })
  price: number;

  constructor(id: string, price: number) {
    this.id = id;
    this.price = price;
  }
}
```

Here is the helper that generates a dynamic class based on the table name, and created the EntitySchema

```
export const generateCardPriceEntity = (versionTitle: string) => {
  const EntityClass = class extends CardPrice {
    constructor(id: string, price: number) {
      super(id, price);
    }
  };

  const schema = new EntitySchema({
    tableName: versionTitle,
    name: versionTitle,
    properties: {
      id: { primary: true, type: 'string' },
      price: { type: 'number', nullable: true },
    },
  });

  return { schema, EntityClass };
};
```

and here is my service code that is trying to register the entity and create the rows in the new table. Notice Im not actually doing anything with the EntitySchema. 

```
  async populateCardPriceTable(
    versionTitle: string,
    priceData: { id: string; price: number }[],
  ) {
  const { schema: _, EntityClass } = generateCardPriceEntity(versionTitle);
  await this.orm.discoverEntity(EntityClass);

  for (const entity of priceData) {
    await this.em.create(versionTitle, entity);
  }
}
```

This version of the code cannot find the entity metadata. 
```
MetadataError: Metadata for entity MyCustomEntity not found
```

I have also tried a version where I used `Object.defineProperty` in the `generateCardPriceEntity` helper to define the class `name` property
```Object.defineProperty(EntityClass, 'name', { value: versionTitle });```

this version runs into the same postgres error I posted above. It gets past the calls the `em.create`, but when actually persisting the entities, it errors out.

I'm trying to use the approach you are describing, but it seems like I'm missing something? Thanks in advance! <3

**Reply by B4nan** [reply]
You are mixing all the approaches, I told you not to use a class at all, but you still have one, and still with decorators. Just remove that part of your code, and use the string name in `em.create`. You should feed the EntitySchema instance to `discoverEntity`, I see its type is not allowing that, will fix that in a minute. Also, `em.create` is a sync method that only creates the entity instance.

```ts
export const generateCardPriceEntity = (versionTitle: string) => {
  const schema = new EntitySchema({
    tableName: versionTitle,
    name: versionTitle,
    properties: {
      id: { primary: true, type: 'string' },
      price: { type: 'number', nullable: true },
    },
  });

  return schema;
};

async populateCardPriceTable(
    versionTitle: string,
    priceData: { id: string; price: number }[],
) {
  const schema = generateCardPriceEntity(versionTitle);
  await this.orm.discoverEntity(schema as any); // type is wrong here

  for (const entity of priceData) {
    this.em.create(versionTitle, entity);
  }

  await this.em.flush();
}
```

It also feels a bit like you plan to do this repetitively - you need to discover such entity only once, afterwards you can refer to it via the string name, and you can access its metadata, including the dynamic class reference, e.g. via `orm.getMetadata(name)`. So if something, there should be some `ensureEntityMetadata` method, you are not supposed to be rediscovering it again and again.

**Reply by ZakRabe** [reply]
> You should feed the EntitySchema instance to `discoverEntity`, I see its type is not allowing that, will fix that in a minute. Also, `em.create` is a sync method that only creates the entity instance.

Ahh, this makes sense. Thats why I was using a class still, I assumed I needed it to pass to `discoverEntity` since it was typed as `Constructor`, thanks for the clarity here!

> It also feels a bit like you plan to do this repetitively - you need to discover such entity only once, afterwards you can refer to it via the string name, and you can access its metadata, including the dynamic class reference, e.g. via `orm.getMetadata(name)`. So if something, there should be some `ensureEntityMetadata` method, you are not supposed to be rediscovering it again and again.

Correct, I will be creating multiple tables this way. My plan was to discover existing entities in the system at startup. I didn't show it in my code examples, but I also create rows in another table for each copy. So at system startup I can easily manually discover existing copies of the table using the schema approach.  I'll only be discovering during runtime when a new copy is created by the user.

I have updated my code to match your approach, but I have still been hitting errors. mostly related to undefined metadata.

The first few errors were missing things like `hydrateProps` and/or `comparableProps`. I tried defining these in the schema, but I encountered new similar errors down the line. 

for example: this kind of errors happened with hydrateProps, comparableProps, trackingProps. I would define them as best I could in the schema definition.
```
@mikro-orm\core\entity\EntityHelper.js:239
        for (const prop of meta.trackingProps) {
                                ^

TypeError: meta.trackingProps is not iterable
```


I figured I would upgrade my project to use the latest version of mikro-orm. I thought maybe these were issues in the `discoverEntities` method, or the `EntitySchema` constructor, that had been resolved in a later release. I was able to bump to the latest version: 6.3.9.

I am hitting similar undefined metadata errors from the `EntityFactory` now:
```
@mikro-orm\core\entity\EntityHelper.js:245
        Object.defineProperties(entity, meta.definedProperties);
               ^

TypeError: Cannot convert undefined or null to object
    at Function.defineProperties (<anonymous>)
    at EntityHelper.ensurePropagation (@mikro-orm\core\entity\EntityHelper.js:245:16)   
    at EntityFactory.createEntity (@mikro-orm\core\entity\EntityFactory.js:209:45)      
    at EntityFactory.create (@mikro-orm\core\entity\EntityFactory.js:58:39)
    at SqlEntityManager.create (@mikro-orm\core\EntityManager.js:1183:41)
    at CardService.populateCardPriceTable (src\card\card.service.ts:60:15)
```
    
I am not familiar enough with the implementation or expectations of the ORM system to deduce if this is a bug, or my own mistake.   
These other metadata attribute definitions require a lot more data by the type definition. I have also tried using the instance methods on the schema, but the error persists. 

```export const generateCardPriceSchema = (versionTitle: string) => {
  const schema = new EntitySchema({
    tableName: versionTitle,
    name: versionTitle,

    // properties: {
    //   id: { primary: true, type: 'string' },
    //   price: { type: 'number', nullable: true },
    // },
  });

  schema.addPrimaryKey('id', 'string');
  schema.addProperty('price', 'number', { nullable: true });

  return schema;
};
```

**Reply by B4nan** [reply]
> The first few errors were missing things like hydrateProps and/or comparableProps. 

This is exactly what the `discoverEntity` method should handle for you.

Let me take a closer look, will add some test when I fix the type issue I pointed above to see if it actually works.

**Reply by ZakRabe** [reply]
Thanks very much! 

I was able to get past the metadata errors by copying the properties into these two arrays.

```
export const generateCardPriceSchema = (versionTitle: string) => {
  const schema = new EntitySchema({
    tableName: versionTitle,
    name: versionTitle,
    definedProperties: {
      id: { primary: true, type: 'string' },
      price: { type: 'number', nullable: true },
    },
    hydrateProps: [],
    trackingProps: [],
    // properties: {
    //   id: { primary: true, type: 'string' },
    //   price: { type: 'number', nullable: true },
    // },
  });

  schema.addPrimaryKey('id', 'string');
  schema.addProperty('price', 'number', { nullable: true });
  schema.meta.hydrateProps.push(...Object.values(schema.meta.properties));
  schema.meta.trackingProps.push(...Object.values(schema.meta.properties));

  return schema;
};
```

But after that it looks like the generated hydration code fails on a readonly entity property 
```
JIT runtime error: Cannot assign to read only property 'id' of object '[object Object]'

  function(entity, data, factory, newEntity, convertCustomTypes, schema) {
    if (data.id === null) {
      entity.id = null;
    } else if (typeof data.id !== 'undefined') {
>     entity.id = data.id;
```

If I can do anything to help diagnose or debug the issue, please let me know. thanks for all the help. greatly appreciated!

**Reply by B4nan** [reply]
This should help 3fde7d42a6bc2cac4374d07d22019135fa4963d9

**Reply by ZakRabe** [reply]
I'll keep a lookout for the next release and give it a try. 
I'll post my final solution here for future readers, and I'll submit a docs PR for the EntitySchema page. Thanks very much!

**Comment by ZakRabe** [comment, marked as answer]
After bumping to the latest version, this is working great. Thanks for the quick fix @B4nan!

For anyone reading in the future, the solution I went with was creating an entity class that represented my "base table", this table could be created via migrations, but wasn't actually where I would persist my data. 

My base table and its clones only have a primary key and one column. This model is unlikely to change in the future, so not being able to generate migrations for the generated tables in the future was not a worry for me.

When I create a new copy of the table, I create a row in another table that records the table name. When my service starts, it looks up the existing rows in that table, and creates a new `EntitySchema` for each, and register those via `discoverEntities`.

When creating a new table during my application runtime: I copied that base table at runtime using `em.execute`, then I used `EntitySchema` and `discoverEntity` to make the ORM aware of it. 

Here is the relevant code for creating the entity at runtime and registering it in the ORM. 

```
const schema = new EntitySchema({
  tableName: 'myNewTable',
  name: 'myNewTable',
  properties: {
    id: { primary: true, type: 'string' },
    price: { type: 'number', nullable: true, columnType: 'float' },
  },
});
orm.discoverEntity(schema);
em.create('myNewTable', { id: 'abc-123', price: 1.5 });
await em.flush()
```
